### PR TITLE
Normalize Ethereum

### PR DESCRIPTION
**Pull request by matthewjablack**
This PR normalizes date received from Ethereum by implementing `Ethereum Utils`.

### CODE REVIEW

File: src/schema/Block.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
     "hash": {
       "type": "string",
       "title": "Hash",
-      "pattern": "^([A-Fa-f0-9]{64})$"
+      "pattern": "^([0x]{2}?)([A-Fa-f0-9]{64})$"
```
Comment (harshjv) [inline src/schema/Block.json]: Hey @mattBlackDesign, you can remove `0x` from prefix.

File: src/index.js [hunk 2]
```diff
@@ -143,9 +143,12 @@ export default class ChainAbstractionLayer {
 
     const block = await this.provider.getBlockByNumber(blockNumber, includeTx)
 
-    if (!this.validateBlock(block)) {
-      throw new Error('Provider returned an invalid block')
-    }
+    // const valid = this.validateBlock(block)
```
Comment (harshjv) [inline src/index.js]: Why is this in comments? Skipping block test?

File: src/index.js [hunk 3]
```diff
@@ -174,19 +177,19 @@ export default class ChainAbstractionLayer {
   async getTransactionByHash (txHash) {
     this._checkMethod('getTransactionByHash')
 
-    if (!isString(txHash)) {
-      throw new Error('Transaction hash should be a string')
-    }
+    // if (!isString(txHash)) {
```
Comment (harshjv) [inline src/index.js]: and this too?

**Review by harshjv — CHANGES_REQUESTED** [review]

**Comment by matthewjablack** [comment]
@harshjv this PR allows CAL to work properly with both Bitcoin and Ethereum Multi Chain Block explorer demo. I'll modify this PR to standardize Ethereum to remove 0x at beginning of each attribute

**Comment by matthewjablack** [comment]
@harshjv  I've updated the PR with normalization of data from ethereum chain

File: src/index.js [hunk 4]
```diff
@@ -184,7 +187,10 @@ export default class ChainAbstractionLayer {
 
     const transaction = await this.provider.getTransactionByHash(txHash)
 
-    if (!this.validateTransaction(transaction)) {
+    const valid = this.validateTransaction(transaction)
+
+    if (!valid) {
+      console.log(this.validateTransaction.errors)
```
Comment (monokh) [inline src/index.js]: probably better to include this in the error you are throwing? Will still get it logged + the metadata of the error is also carried through

File: src/providers/ethereum/EthereumUtil.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import BlockSchema from '../../schema/Block.json'
+import TransactionSchema from '../../schema/Transaction.json'
+
+function formatEthHex (match, p1, p2, offset, string) {
```
Comment (monokh) [inline src/providers/ethereum/EthereumUtil.js]: this is quite an odd function 😃 can you describe it?

File: src/providers/ethereum/EthereumUtil.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  return p2
+}
+
+export function formatEthResponse (obj) {
+  if (Array.isArray(obj)) {
+    for (let i = 0; i < obj.length; i++) {
+      if (obj[i].startsWith('0x')) {
```
Comment (monokh) [inline src/providers/ethereum/EthereumUtil.js]: Could this test and replace be just `obj[i].replace('0x', '')` ?

File: src/index.js [hunk 7]
```diff
@@ -143,7 +143,10 @@ export default class ChainAbstractionLayer {
 
     const block = await this.provider.getBlockByNumber(blockNumber, includeTx)
 
-    if (!this.validateBlock(block)) {
+    const valid = this.validateBlock(block)
+
+    if (!valid) {
+      console.log(this.validateBlock.errors)
```
Comment (harshjv) [inline src/index.js]: no debugging related `console.log` in `master` 🙅🏻‍♂️

File: src/providers/ethereum/EthereumUtil.js (hunk 6 above)
Comment (harshjv) [reply src/providers/ethereum/EthereumUtil.js]: Can you use functional programming here? `.map` or `.forEach` for consistency?

**Review by harshjv — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 4 above)
Comment (matthewjablack) [reply src/index.js]: Good idea. Added to errors.

File: src/providers/ethereum/EthereumUtil.js (hunk 5 above)
Comment (matthewjablack) [reply src/providers/ethereum/EthereumUtil.js]: Was using implementation of `.replace` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) but isn't needed anymore 😄

File: src/providers/ethereum/EthereumUtil.js (hunk 6 above)
Comment (matthewjablack) [reply src/providers/ethereum/EthereumUtil.js]: Updated to use `.replace` and using `.map`

File: src/index.js (hunk 7 above)
Comment (matthewjablack) [reply src/index.js]: Added to error instead 😃

**Comment by matthewjablack** [comment]
@monokh  @harshjv  updated with your changes

**Review by harshjv — APPROVED** [review]
LGTM. In next sprint, we will need to figure out best way of handling errors and formatting inputs/outputs.

**Review by monokh — APPROVED** [review]
Looks good, no conflicts merging

**Pull request merged**